### ISSUE TITLE
Mibi Segmentation Dataloader

### ISSUE THREAD

**Issue by JLrumberger** [angelolab/cell_classification#1]
**Section 1: Design details**

**Relevant background**

To train and validate an ML model that predicts if a given marker is positive or negative for a cell, we need a tfrecords object that stores and returns input and ground truth data as well as meta data (dataset, platform, etc.) of the samples. 

**Design overview**

We'll be using tfrecords for storage and fast i/o during training. For constructing the tfrecords object, we need to construct for each single training example a dictionary that holds all necessary data. For this we develop a number of helper functions that load data from different sources (mibi tifs, dfs holding cell type information, etc.). 

We want our training examples to consist of the following data
- single or multiple channels from multiplexed images (MIBI, CODEX, etc.)
- name of each channel
- name of the dataset
- the imaging platform used to acquire the images
- the cell types as semantic segmentation or as a dictionary
- cell segmentation masks in binary form (either 0/1 => background&cell boundary/cell interior or 0/1/2 => background/cell interior/cell boundary
- ground truth binary segmentation labels for each of the channels

**Design list/flowchart**

The desired output is a tfrecords store that stores all of the above mentioned data. Therefore we need:

1. prepare example class that takes in all required information (list of folder names, path to conversion matrix and cell table, etc.) and serializes the training examples into a tfrecord store. The function will make use of the following helper functions:
2. function that takes an instance segmentation mask and returns a binary mask
3. function that takes a list of cell segmentation indices and returns a list of cell types
4. function that takes a list of cell types and a marker name and returns marker activity as a list that contains 
    - "1" if the marker is positive for the cell type
    - "0" if the marker is negative for the cell type
    - "-1" if the marker is unspecific for the cell type (e.g Vimentin for endothelial cells)
5. function that takes in instance segmentation mask and marker activity list and returns marker activity ground truth mask
6. helper function that iterates through the whole dataset to calculate 99% quantile for each channel for the whole dataset and stores the results in a .json and returns it as a dict

**Required inputs**

- list of folder paths of multiplexed images (height x width) stored as .tiff
- path to conversion matrix (cell types x markers) that holds for each cell type which marker is positive/negative/unspecific
- path to `cell_table.csv` (cells x features) that holds the cell types for each cell
- imaging platform (str)
- dataset (str)
- tile-size [int,int] optional
- key (str) for the cell type column in `cell_table.csv`, default `cluster_labels`
- key (str) for cell segmentation tif, default `cell_segmentation` 

**Output files**

- tfrecord store holding all training samples, consisting of
    - multiplexed image: H x W x C
    - instance segmentation (binary erosion) mask: H x W x 1
    - ground truth mask: H x W x C
    - either instance segmentation mask: H x W x 1 and list of cell types (#cells in ROI) or semantic segmentation mask H x W x 1
    - name of each channel
    - name of the dataset
    - name of the imaging platform
- .json file holding the 99%-quantile for each channel for the whole dataset

**Section 2: Implementation details**

Note to self: store data as .jpeg encoded byte arrays in the tf_record as described in this issue [tf issue 9675](https://github.com/tensorflow/tensorflow/issues/9675).

**Control flow**

Provide additional, more granular details (if necessary) about how the proposed coding logic will be put together

**Milestones and timeline**

List each of the major components of the project, and provide an estimated completion date for each one.

**Comment by ngreenwald** [comment]
Looks great!
